dependencies/spirv/CMakeLists.txt

```diff
@@ -0,0 +1,36 @@
+include(FetchContent)
+include(CMakeDependentOption)
+
+cmake_dependent_option(
+    FETCH_SPIRV_HEADERS "Fetch SPIR-V headers from remote repo (disable and use SPIRV-Headers_SOURCE_DIR to use an existing checkout)." OFF
+    TARGET_SPIRV OFF
+)
+
+if (FETCH_SPIRV_HEADERS)
+    set(SPIRV_VERSION 1.5.4)
+    message(STATUS "Fetching SPIR-V ${SPIRV_VERSION}...")
+    FetchContent_Declare(
+        SPIRV-Headers
+        GIT_REPOSITORY https://github.com/KhronosGroup/SPIRV-Headers.git
+        GIT_TAG ${SPIRV_VERSION}
+        GIT_SHALLOW TRUE
+        GIT_SUBMODULES ""
+    )
+    FetchContent_GetProperties(SPIRV-Headers)
+    if (NOT SPIRV-Headers_POPULATED)
+        FetchContent_Populate(SPIRV-Headers)
+    endif ()
+    set(SPIRV_INCLUDE_DIR ${SPIRV-Headers_SOURCE_DIR}/include)    
+else ()
+    find_path(SPIRV_INCLUDE_DIR spirv/1.0/spirv.h)
+endif ()
+
+if (NOT SPIRV_INCLUDE_DIR)
+    message(FATAL_ERROR "SPIRV-Headers were not found - please enable FETCH_SPIRV_HEADERS or specify a valid path with 'SPIRV_INCLUDE_DIR'!")
+endif ()
+
+message(STATUS "Using SPIR-V headers from ${SPIRV_INCLUDE_DIR}...")
+add_library(Halide_SPIRV INTERFACE IMPORTED)
+target_include_directories(Halide_SPIRV 
+                           SYSTEM # Use -isystem instead of -I; this is a trick so that clang-tidy won't analyze these includes
```

Comment:
[I think](https://cmake.org/cmake/help/latest/prop_tgt/NO_SYSTEM_FROM_IMPORTED.html) `SYSTEM` is unnecessary here since `Halide_SPIRV` is `IMPORTED`. It should also be renamed to `Halide::SPIRV` to ensure that CMake always treats this name as a target (and to therefore be defensive against future typos).

Reply:
I renamed the library target, but left the SYSTEM declaration so that the <> notation for the include file path still worked.

Reply:
Actually, this didn't work, and I reverted back to Halide_SPIRV target names since the dependencies need to exist before Halide itself, and Halide::SPIRV was getting silently ignored for linkage.

Reply:
I would like to repro that since it's not at all what I would expect. Do you have some steps?

Reply:
Just changing the target name from `Halide_SPIRV` to `Halide::SPIRV` like this:

`add_library(Halide::SPIRV INTERFACE)`
`target_include_directories(Halide::SPIRV SYSTEM INTERFACE "${SPIRV_INCLUDE_DIR}")`

Causes errors with cmake v3.23.2:

`CMake Error at dependencies/spirv/CMakeLists.txt:35 (add_library):
  The target name "Halide::SPIRV" is reserved or not valid for certain CMake
  features, such as generator expressions, and may result in undefined
  behavior.

CMake Error at dependencies/spirv/CMakeLists.txt:36 (target_include_directories):
  Cannot specify include directories for target "Halide::SPIRV" which is not
  built by this project.
`

Reply:
Using the old target name and adding an ALIAS for Halide::SPIRV like so:

````
add_library(Halide_SPIRV INTERFACE)
add_library(Halide::SPIRV ALIAS Halide_SPIRV)
target_include_directories(Halide_SPIRV SYSTEM INTERFACE "${SPIRV_INCLUDE_DIR}")
````

And used like before:

````
if (TARGET Halide::SPIRV)
    target_link_libraries(Halide PRIVATE Halide::SPIRV)
endif()

````
DOES add the correct -isystem path to the compiler definitions ... but then we incur a dependency on Halide_TARGETS for the interface library Halide::SPIRV which isn't installed.

Reply:
You would have to change from:
```
add_library(Halide_SPIRV INTERFACE)
```
to
```
add_library(Halide::SPIRV INTERFACE IMPORTED)
```
since names with colons `::` can only be used in `IMPORTED` and `ALIAS` targets. When a target is `IMPORTED` its `INTERFACE_INCLUDE_DIRECTORIES` property (which is set by `target_include_directories(... INTERFACE ...)`) is treated as `SYSTEM` by default.

> but then we incur a dependency on Halide_TARGETS for the interface library Halide::SPIRV which isn't installed

Since `Halide::SPIRV` is meant to be _totally_ private, just write
```
target_link_libraries(Halide PRIVATE "$<BUILD_INTERFACE:Halide::SPIRV>")
```
instead.

Reply:
Cool ... thanks for the detailed explanation!